src/v/test_utils/http_imposter.cc

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "test_utils/http_imposter.h"
+
+#include "config/node_config.h"
+#include "vlog.h"
+
+#include <seastar/http/function_handlers.hh>
+
+#include <utility>
+
+static ss::logger http_imposter_log("http_imposter"); // NOLINT
+
+http_imposter_fixture::http_imposter_fixture() {
+    _server = ss::httpd::http_server_control();
+    _server.start().get();
+    ss::ipv4_addr ip_addr = {httpd_host_name.data(), httpd_port_number};
+    _server_addr = ss::socket_address(ip_addr);
+}
+
+http_imposter_fixture::~http_imposter_fixture() { _server.stop().get(); }
+
+const std::vector<ss::httpd::request>&
+http_imposter_fixture::get_requests() const {
+    return _requests;
+}
+
+const std::multimap<ss::sstring, ss::httpd::request>&
+http_imposter_fixture::get_targets() const {
+    return _targets;
+}
+
+void http_imposter_fixture::listen() {
+    _server.set_routes([this](ss::httpd::routes& r) { set_routes(r, _urls); })
+      .get();
+    _server.listen(_server_addr).get();
+}
+
+static ss::sstring remove_query_params(std::string_view url) {
+    auto q_pos = url.find('?');
+    return q_pos == std::string_view::npos ? ss::sstring{url.data(), url.size()}
+                                           : ss::sstring{url.substr(0, q_pos)};
+}
+
+struct content_handler {
+    content_handler(
+      const http_test_utils::registered_urls& urls, http_imposter_fixture& imp)
+      : _registered_urls{urls}
+      , fixture(imp) {}
+
+    ss::sstring
+    handle(ss::httpd::const_req request, ss::httpd::reply& repl) const {
+        fixture.requests().push_back(request);
+        fixture.targets().insert(std::make_pair(request._url, request));
+
+        vlog(
+          http_imposter_log.trace,
+          "S3 imposter request {} - {} - {}",
+          request._url,
+          request.content_length,
+          request._method);
+
+        ss::httpd::request lookup_r{request};
+        lookup_r._url = remove_query_params(request._url);
+
+        auto response = _registered_urls.lookup(lookup_r);
+        repl.set_status(response.status);
+        return response.body;
+    }
+
+    const http_test_utils::registered_urls& _registered_urls;
+    http_imposter_fixture& fixture;
+};
+
+void http_imposter_fixture::set_routes(
+  ss::httpd::routes& r, const http_test_utils::registered_urls& urls) {
+    using namespace ss::httpd;
+    _handler = std::make_unique<function_handler>(
+      [urls, this](const_req req, reply& repl) {
```

Comment:
I believe `urls` is being copied here, which seems odd because its really `this::_urls`. Since the `content_handler` has a reference to `this` it can probably obtain urls itself

Reply:
good point, I think this got mixed up extracting code from multiple places, I've removed the urls being passed around and code is much cleaner now, thanks